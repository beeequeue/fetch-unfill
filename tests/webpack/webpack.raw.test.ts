import path from "node:path"

import { createFsFromVolume, Volume } from "memfs"
import { describe, expect, it } from "vitest"
import { type Configuration, optimize, type OutputFileSystem, webpack } from "webpack"
import { merge } from "webpack-merge"

import { createTester, type Output } from "../utils.js"

const test = createTester(
  "webpack",
  async (name: string, extraOptions: Configuration | null = {}) => {
    const compiler = webpack(
      merge(
        {
          entry: {
            cjs: path.resolve(__dirname, "..", "fixtures", `${name}.cjs`),
            mjs: path.resolve(__dirname, "..", "fixtures", `${name}.mjs`),
          },
          output: { libraryTarget: "commonjs-module" },
          plugins: [
            new optimize.LimitChunkCountPlugin({
              maxChunks: 1, // disable creating additional chunks
            }),
          ],

          mode: "production",
          target: "node18",
          stats: "none",
          devtool: false,
          experiments: { futureDefaults: true },
        },
        extraOptions ?? {},
      ),
    )

    const output = createFsFromVolume(new Volume())
    compiler.outputFileSystem = output as OutputFileSystem

    const result = await new Promise<Output | Error>((resolve, reject) =>
      compiler.run((err, stats) => {
        if (err != null || stats?.hasErrors()) {
          return reject(err ?? stats!.toJson({ errors: true }).errors?.[0])
        }

        resolve({
          cjs: output.readFileSync("dist/cjs.js", { encoding: "utf8" }) as string,
          mjs: output.readFileSync("dist/mjs.js", { encoding: "utf8" }) as string,
        })
      }),
    )

    expect(result).not.toBeInstanceOf(Error)

    return result as Output
  },
)

const alias = {
  resolve: {
    alias: {
      "node-fetch": "fetch-unfill",
      "cross-fetch": "fetch-unfill",
    },
  },
} satisfies Configuration

describe("node-fetch", () => {
  it("bundles the original package", async () => {
    await test("node-fetch", null)
  })

  it("unfills it", async () => {
    await test("node-fetch", alias, true)
  })
})

describe("cross-fetch", () => {
  it("bundles the original package", async () => {
    await test("cross-fetch", null)
  })

  it("unfills it", async () => {
    await test("cross-fetch", alias, true)
  })
})
