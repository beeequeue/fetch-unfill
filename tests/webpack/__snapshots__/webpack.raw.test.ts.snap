// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`cross-fetch > bundles the original package > cjs 1`] = `"301.95KB"`;

exports[`cross-fetch > bundles the original package > mjs 1`] = `"301.98KB"`;

exports[`cross-fetch > unfills it > cjs 1`] = `"529B"`;

exports[`cross-fetch > unfills it > mjs 1`] = `"614B"`;

exports[`cross-fetch > unfills it > unfilled cjs code 1`] = `"(()=>{var e={65:(e,o,s)=>{const t=s(883);e.exports=()=>t("https://example.com"),e.exports.example=()=>t("https://example.com")},883:e=>{const o=globalThis.fetch;globalThis.fetch.polyfill=!1,o.polyfill=!1,e.exports=o,Object.assign(e.exports,{fetch:o,Request:globalThis.Request,Response:globalThis.Response,Headers:globalThis.Headers,AbortController:globalThis.AbortController})}},o={},s=function s(t){var l=o[t];if(void 0!==l)return l.exports;var r=o[t]={exports:{}};return e[t](r,r.exports,s),r.exports}(65);module.exports=s})();"`;

exports[`cross-fetch > unfills it > unfilled mjs code 1`] = `"(()=>{"use strict";var e={d:(o,t)=>{for(var l in t)e.o(t,l)&&!e.o(o,l)&&Object.defineProperty(o,l,{enumerable:!0,get:t[l]})},o:(e,o)=>Object.prototype.hasOwnProperty.call(e,o),r:e=>{"undefined"!=typeof Symbol&&Symbol.toStringTag&&Object.defineProperty(e,Symbol.toStringTag,{value:"Module"}),Object.defineProperty(e,"__esModule",{value:!0})}},o={};e.r(o),e.d(o,{default:()=>n,example:()=>i});const t=globalThis.fetch;globalThis.fetch.polyfill=!1,t.polyfill=!1;const l=t,{Request:r,Response:a,Headers:p,AbortController:s}=globalThis,n=()=>l("https://example.com"),i=()=>l("https://example.com");module.exports=o})();"`;

exports[`node-fetch > bundles the original package > cjs 1`] = `"88.35KB"`;

exports[`node-fetch > bundles the original package > mjs 1`] = `"85.62KB"`;

exports[`node-fetch > unfills it > cjs 1`] = `"531B"`;

exports[`node-fetch > unfills it > mjs 1`] = `"614B"`;

exports[`node-fetch > unfills it > unfilled cjs code 1`] = `"(()=>{var e={533:(e,o,s)=>{const t=s(883);e.exports=()=>t("https://example.com"),e.exports.example=()=>t("https://example.com")},883:e=>{const o=globalThis.fetch;globalThis.fetch.polyfill=!1,o.polyfill=!1,e.exports=o,Object.assign(e.exports,{fetch:o,Request:globalThis.Request,Response:globalThis.Response,Headers:globalThis.Headers,AbortController:globalThis.AbortController})}},o={},s=function s(t){var l=o[t];if(void 0!==l)return l.exports;var r=o[t]={exports:{}};return e[t](r,r.exports,s),r.exports}(533);module.exports=s})();"`;

exports[`node-fetch > unfills it > unfilled mjs code 1`] = `"(()=>{"use strict";var e={d:(o,t)=>{for(var l in t)e.o(t,l)&&!e.o(o,l)&&Object.defineProperty(o,l,{enumerable:!0,get:t[l]})},o:(e,o)=>Object.prototype.hasOwnProperty.call(e,o),r:e=>{"undefined"!=typeof Symbol&&Symbol.toStringTag&&Object.defineProperty(e,Symbol.toStringTag,{value:"Module"}),Object.defineProperty(e,"__esModule",{value:!0})}},o={};e.r(o),e.d(o,{default:()=>n,example:()=>i});const t=globalThis.fetch;globalThis.fetch.polyfill=!1,t.polyfill=!1;const l=t,{Request:r,Response:a,Headers:p,AbortController:s}=globalThis,n=()=>l("https://example.com"),i=()=>l("https://example.com");module.exports=o})();"`;
